clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java

```diff
@@ -264,8 +273,11 @@ public boolean hasValidVersion() {
         return oldestVersion() <= latestVersion();
     }
 
-    public Optional<ApiVersionsResponseData.ApiVersion> toApiVersion(boolean enableUnstableLastVersion) {
-        short oldestVersion = oldestVersion();
+    public Optional<ApiVersionsResponseData.ApiVersion> toApiVersion(boolean enableUnstableLastVersion,
+                                                                     Optional<ApiMessageType.ListenerType> listenerType) {
+        // see `PRODUCE_MIN_VERSION` for details on why we do this
+        short oldestVersion = (this == PRODUCE && listenerType.map(l -> l == ApiMessageType.ListenerType.BROKER).orElse(false)) ?
```

Comment:
Excuse me, but why change the (produce) schema version after we decided to change the API response? I assumed all we needed to do was return the API response including v0-v2 produce requests to fix the librdkafka issue.  Then, the schema could remain at v3-v12, so the v0-v2 requests would still be rejected.

Reply:
I will explain the changes in more detail after I confirm that the test results are passing.

Reply:
We're discussing this in the main discussion, so will resolve this.